### PR TITLE
[Fixes #7085] ApplicationModelConventionExtensions should make a copy…

### LINKED ISSUE #7085: ApplicationModelConventionExtensions should make a copy of collections when iterating them
If you try to add/remove actions from inside a convention, you'll be greated by a fun error like the following:

```
InvalidOperationException: Collection was modified; enumeration operation may not execute.
System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion()
System.Collections.Generic.List+Enumerator.MoveNextRare()
Microsoft.Extensions.DependencyInjection.ApplicationModelConventionExtensions+ActionApplicationModelConvention.Apply(ApplicationModel application)
Microsoft.AspNetCore.Mvc.Internal.ApplicationModelConventions.ApplyConventions(ApplicationModel applicationModel, IEnumerable<IApplicationModelConvention> conventions)
Microsoft.AspNetCore.Mvc.Internal.ControllerActionDescriptorProvider.GetDescriptors()
Microsoft.AspNetCore.Mvc.Internal.ControllerActionDescriptorProvider.OnProvidersExecuting(ActionDescriptorProviderContext context)
Microsoft.AspNetCore.Mvc.Internal.ActionDescriptorCollectionProvider.UpdateCollection()
Microsoft.AspNetCore.Mvc.Internal.ActionDescriptorCollectionProvider.get_ActionDescriptors()
Microsoft.AspNetCore.Mvc.Internal.AttributeRoute.GetTreeRouter()
Microsoft.AspNetCore.Mvc.Internal.AttributeRoute.RouteAsync(RouteContext context)
Microsoft.AspNetCore.Routing.RouteCollection+<RouteAsync>d__9.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Builder.RouterMiddleware+<Invoke>d__4.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware+<Invoke>d__7.MoveNext()
```

We should copy these collections before iterating them so you don't get errors like this.

### PR DESCRIPTION
**Pull request by kichalla**
… of collections when iterating them

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs [hunk 1]
```diff
@@ -148,11 +149,16 @@ public void Apply(ApplicationModel application)
                     throw new ArgumentNullException(nameof(application));
                 }
 
-                foreach (var controller in application.Controllers)
+                // Create copies of collections of controllers, actions and parameters as users could modify
+                // these collections from within the convention itself.
+                var controllers = application.Controllers.ToArray();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs]: Guessing you also looked at Razor Pages conventions to make sure it doesn't need something like this

File: test/Microsoft.AspNetCore.Mvc.Core.Test/DependencyInjection/ApplicationModelConventionExtensionsTest.cs [hunk 2]
```diff
@@ -182,5 +264,37 @@ public void Apply(ControllerModel controller)
...
+            {
+                application.Controllers.RemoveAt(0);
+            }
+        }
+
+        private class ActionModelCollectionModifyingConvention : IActionModelConvention
+        {
+            public void Apply(ActionModel action)
+            {
+                action.Controller.Actions.Remove(action);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/DependencyInjection/ApplicationModelConventionExtensionsTest.cs]: Does this work well if you add an action? We could possibly support it (if we indexed in to the collection and tracked if it's still there to track removals). Would be good to find out more from @rynowak what the exact scenario was.

File: src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs]: Yeah I think pages also needs this too, it has parameters at least.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/DependencyInjection/ApplicationModelConventionExtensionsTest.cs (hunk 2 above)
Comment (rynowak) [reply test/Microsoft.AspNetCore.Mvc.Core.Test/DependencyInjection/ApplicationModelConventionExtensionsTest.cs]: Nothing complicated is needed here. The user code sees the original list.

File: src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs (hunk 1 above)
Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs]: Checked with @pranavkm regarding this. We do not have similar concerns with conventions in RazorPages. https://github.com/aspnet/Mvc/issues/6935 issue is for handling parameters and has not been implemented yet.

**Comment by rynowak** [comment]
:shipit: then 👍

**Pull request merged**